File: leetCode/28.实现strStr().cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        if(needle.empty())return 0;
        int i=0,m=haystack.size();//文本指针i用来指示文本串
        int j=0,n=needle.size();//模式指针j用来指示模式串
        vector<int> nextVal=get_nextVal(needle);
        while(i<m&&j<n){
            if(j==-1||needle[j]==haystack[i])i++,j++;
            else j=nextVal[j];
        }
        return j==n?i-j:-1;
    }

    //普通：get_next
    vector<int> get_next(const string& pat){
        int n=pat.size();
        vector<int> next(n+1,-1);//这里数组大小为n+1，要算上pat[n-1]
        int k=-1,j=0;
        while(j<n){
            if(k==-1||pat[k]==pat[j])next[++j]=++k;//k，j指针都右移一位
            else k=next[k];//后缀指针k回退到next[k]（下标）的位置
        }
        return next;
    }

    //优化：get_nextVal
    vector<int> get_nextVal(const string& pat){
        int n=pat.size();
        vector<int> nextVal(n,-1);//这里next数组大小为n，已经算上pat[n-1]了
        int k=-1,j=0;
        while(j<n-1){
            if(k==-1||pat[k]==pat[j]){
                k++,j++;
                if(pat[j]!=pat[k])nextVal[j]=k;
                else nextVal[j]=nextVal[k];
            }
            else k=nextVal[k];
        }
        return nextVal;
    }

};

File: leetCode/28. 实现 strStr().js
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function (haystack, needle) {
  if (needle === "") {
    return 0;
  }
  let i = 0,m = haystack.length;
  let j = 0,n = needle.length;
  let next = getNext(needle)
  while(i<m && j<n){
    if(j == -1 || needle[j] === haystack[i]){
      i++;j++;
    }
    else{
      j = next[j]
    }
  }
  return j===n?i-j:-1
};

function getNext(needle) {
  let next = [];
  next[0] = -1;
  let k = -1,j = 0;
  while(j<needle.length){
    if(k == -1 || needle[k] === needle[j]){
      next[++j] = ++k;
    }
    else{
      k = next[k]; //后缀指针k回退到next[k]的位置
    }
  }
  return next;
}

//优化getNextVal
function getNextVal(needle){
  let nextVal = []
  nextVal[0] = -1
  let k = -1,j = 0;
  while(j<n-1){
    if(k === -1 || needle[k] === needle[j]){
      k++;j++;
      if(needle[k] !== needle[j]){
        nextVal[j] = k;
      }
      else{
        nextVal[j] = nextVal[k]
      }
    }
    else{
      k = next[k]
    }
  }
  return nextVal
}
